src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py

```diff
@@ -0,0 +1,451 @@
+    '''Utility function that takes Fortran code that is unsupported by the
+    DotProduct2CodeTrans transformation and checks that the expected
+    exception is raised when the validate function is applied.
+
+    :param str code: the tangent linear code stored in a string.
+    :param str expected: the expected exception text.
+    :param fortran_reader: converts Fortran into PSyIR.
+    :type fortran_reader: :py:class:`psyclone.psyir.frontend.fortran`
+    :param integer index: specifies which binaryoperation to check in \
+        code order. Defaults to 0 (which is the first).
+
+    '''
+    psyir = fortran_reader.psyir_from_source(code)
+    dot_product = psyir.walk(BinaryOperation)[index]
+    assert dot_product.operator == BinaryOperation.Operator.DOT_PRODUCT
+    trans = DotProduct2CodeTrans()
+    with pytest.raises(TransformationError) as info:
+        trans.validate(dot_product)
+    assert expected in str(info.value)
+
+
+def check_trans(code, expected, fortran_reader, fortran_writer, tmpdir,
+                index=0):
+    '''Utility function that takes fortran code and checks that the
+    expected code is produced when the DotProduct2CodeTrans
+    transformation is applied. The code is also checked to see if it
+    will compile.
+
+    :param str code: the tangent linear code stored in a string.
+    :param str expected: the expected adjoint code stored in a string.
+    :param fortran_reader: converts Fortran into PSyIR.
+    :type fortran_reader: :py:class:`psyclone.psyir.frontend.fortran`
+    :param fortran_writer: converts PSyIR into Fortran.
+    :type fortran_writer: :py:class:`psyclone.psyir.backend.fortran`
+    :param tmpdir: path to a test-specific temporary directory in \
+        which to test compilation.
+    :type tmpdir: :py:class:`py._path.local.LocalPath`
+    :param integer index: specifies which binaryoperation to check in \
+        code order. Defaults to 0 (which is the first).
+
+    '''
+    psyir = fortran_reader.psyir_from_source(code)
+    dot_product = psyir.walk(BinaryOperation)[index]
+    assert dot_product.operator == BinaryOperation.Operator.DOT_PRODUCT
+    trans = DotProduct2CodeTrans()
+    trans.apply(dot_product)
+    result = fortran_writer(psyir)
+    assert expected in result
+    assert Compile(tmpdir).string_compiles(result)
+
+
+# _get_array_bound function
+
+def test_bound_error(fortran_writer):
+    '''Test that the expected exception is raised if both arguments are
+    not references to arrays.
+
+    '''
+    with pytest.raises(TransformationError) as info:
+        _get_array_bound(Literal("1.0", REAL_TYPE), Literal("2.0", REAL_TYPE),
+                         fortran_writer)
+    assert ("DotProduct2CodeTrans._get_array_bound requires at least one of "
+            "the dotproduct arguments to be an array but found '1.0' and "
+            "'2.0'." in str(info.value))
+
+
+@pytest.mark.parametrize("dim1,dim2", [("2:10", "2:10"), (":", "2:10"),
```

Comment:
What about a non-unit stride?

Reply:
The input is assumed to be unit stride and this is a constraint in the validate method so non-unit stride will never be passed to this function. I've added some words to the docstring of the function. However, I admit that the constraint in the validate method is also that only full ranges are supported so the functionality here is beyond what is required (but that does not break anything).